README.md
<h1 align="center"><strong>Fareview</strong></h1>

<p align="center">
  <img width="auto" height="auto" src="https://media.giphy.com/media/3o6MbtelsDZdsbFB7i/giphy.gif">
</p>
<br />

[![CI](https://github.com/ngshiheng/fareview/actions/workflows/ci.yml/badge.svg)](https://github.com/ngshiheng/fareview/actions/workflows/ci.yml)
[![CD](https://github.com/ngshiheng/fareview/actions/workflows/cd.yml/badge.svg)](https://github.com/ngshiheng/fareview/actions/workflows/cd.yml)
[![Generate CSV](https://github.com/ngshiheng/fareview/actions/workflows/generate_csv.yml/badge.svg)](https://github.com/ngshiheng/fareview/actions/workflows/generate_csv.yml)

## What is Fareview?

✔️ A simple & easy to use e-commerce price monitoring tool for vendors.

🖥 Monitor product prices on top e-commerce (shopee.sg, lazada.sg, qoo10.sg) platforms in Singapore.

📈 Increase your sales and profit margins by uncovering competitors' prices.

## Table of Contents

- [What is Fareview?](#what-is-fareview)
- [Table of Contents](#table-of-contents)
- [Disclaimer](#disclaimer)
- [Features](#features)
- [Content](#content)
- [Development Setup](#development-setup)
  - [Installation](#installation)
  - [Pre-commit Hooks](#pre-commit-hooks)
  - [Database](#database)
- [Usage](#usage)
  - [Start crawling](#start-crawling)
    - [Run single spider](#run-single-spider)
    - [Run all spiders](#run-all-spiders)
    - [Run all spiders, in parallel](#run-all-spiders-in-parallel)
  - [Proxy & Sentry (optional)](#proxy--sentry-optional)
- [Contributing](#contributing)

## Disclaimer

This software is only used for research purposes, users must abide by the relevant laws and regulations of their location, please do not use it for illegal purposes. The user shall bear all the consequences caused by illegal use.

## Features

-   [x] Automated random user agent rotation
-   [x] Colored logging
-   [x] CSV update [cron job](./.github/workflows/generate_csv.yml) that updates [data](./data/)
-   [x] Data deduplication pipeline
-   [x] Database migration with [Alembic](https://alembic.sqlalchemy.org/en/latest/)
-   [x] Delayed requests middleware
-   [x] Error monitoring and alerting with [Sentry](https://sentry.io/)
-   [x] Proxied requests with [Scraper API](https://www.scraperapi.com/?fp_ref=jerryng)
-   [x] Requests retry
-   [x] Uses [Railway](https://railway.app?referralCode=jerrynsh) PostgreSQL

## Content

-   You may find the list of commercial beer prices under [`data/`](./data/) or on [this Google Sheets](https://s.jerrynsh.com/fareview)
-   The prices are updated **once daily** using [GitHub action](./.github/workflows/generate_csv.yml)

## Development Setup

### Installation

Make sure you have [poetry](https://python-poetry.org/docs/#installation) installed on your machine.

```sh
poetry install

# Installing dependencies only
poetry install --no-root

# Updating dependencies to their latest versions
poetry update
```

### Pre-commit Hooks

Before you begin your development work, make sure you have installed [pre-commit hooks](https://pre-commit.com/index.html#installation).

Some example useful invocations:

-   `pre-commit install`: Default invocation. Installs the pre-commit script alongside any existing git hooks.
-   `pre-commit install --install-hooks --overwrite`: Idempotently replaces existing git hook scripts with pre-commit, and also installs hook environments.

### Database

-   Make sure you have a running instance of the latest PostgreSQL in your local machine.

```sh
# Example to spin up a PostgreSQL Docker instance locally
docker run -d --name dpostgres -p 5432:5432 -e POSTGRES_HOST_AUTH_METHOD=trust postgres:latest
```

-   By default, the database for this project should be named as `fareview`.
-   For database migration steps, please read [this](alembic/README.md).

## Usage

### Start crawling

#### Run single spider

```sh
# To list all spiders
poetry run scrapy crawl list

# To run a single spider
poetry run scrapy crawl shopee

# To run single spider with json output
poetry run scrapy crawl shopee -o shopee.json
```

#### Run all spiders

```sh
poetry run scrapy list | xargs -n 1 poetry run scrapy crawl
```

#### Run all spiders, in parallel

```sh
scrapy list | xargs -n 1 -P 0 scrapy crawl
```

### Proxy & Sentry (optional)

[ScraperAPI](https://www.scraperapi.com/?fp_ref=jerryng) is used as our proxy server provider. [Sentry](https://sentry.io/) is used for error monitoring.

```sh
export SCRAPER_API_KEY="<YOUR_SCRAPER_API_KEY>"
export SENTRY_DSN="<YOUR_SENTRY_DSN>"
```

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

1. Fork this
2. Create your feature branch (`git checkout -b feature/bar`)
3. Please make sure you have installed the `pre-commit` hook and make sure it passes all the lint and format check
4. Commit your changes (`git commit -am 'feat: add some bar'`, make sure that your commits are [semantic](https://www.conventionalcommits.org/en/v1.0.0/#summary))
5. Push to the branch (`git push origin feature/bar`)
6. Create a new Pull Request


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.2.2 (2022-12-31)
### Fix
* **deps:** Update dependency sqlalchemy to v1.4.45 ([`7a0e6a2`](https://github.com/ngshiheng/fareview/commit/7a0e6a2c01d04c266334f70c2c80eea327a82783))

## v1.2.1 (2022-12-21)
### Fix
* **spider:** Increase download delay & add delay between requests for shopee ([`0b69cae`](https://github.com/ngshiheng/fareview/commit/0b69caef2edf20345adf7bb8f6ad0eb9aae203b8))

## v1.2.0 (2022-12-18)
### Feature
* **spider:** Handle captcha challenge ([`c6ce456`](https://github.com/ngshiheng/fareview/commit/c6ce4560e26d338c6454064f792ce343a8738a69))

## v1.1.39 (2022-12-16)
### Fix
* **spider:** Handle captcha ([`71a090e`](https://github.com/ngshiheng/fareview/commit/71a090e8e9b7f21f03dbbac6f51e76748f55012a))

## v1.1.38 (2022-12-06)
### Fix
* **deps:** Pin dependency setuptools to 65.6.3 ([`d7da75e`](https://github.com/ngshiheng/fareview/commit/d7da75ee53d6d12754eda0929668fd608bf7663b))

## v1.1.37 (2022-12-05)
### Fix
* Poetry tries to remove setuptools and then complains it can't find it ([`acde39e`](https://github.com/ngshiheng/fareview/commit/acde39eef170d899a5a4afa9a5ba4b2f24709f97))
* **deps:** Update dependency scrapy to v2.7.1 ([`470a193`](https://github.com/ngshiheng/fareview/commit/470a193260d1f6889b059ca75333cfbcf5c394f1))

## v1.1.36 (2022-12-03)
### Fix
* **deps:** Update dependency sentry-sdk to v1.11.1 ([`37bb386`](https://github.com/ngshiheng/fareview/commit/37bb3862f8de3401b16cf9026359d4d11fc49fda))

## v1.1.35 (2022-12-03)
### Fix
* **deps:** Update dependency colorlog to v6.7.0 ([`2207fc0`](https://github.com/ngshiheng/fareview/commit/2207fc0484fd856286f712843b9b717a7120713d))

## v1.1.34 (2022-12-03)
### Fix
* **deps:** Update dependency ipython to v8.7.0 ([`69eabfa`](https://github.com/ngshiheng/fareview/commit/69eabfa8bf44d62434f62e86835df8745b5f10a1))

## v1.1.33 (2022-11-23)
### Fix
* Install setuptool ([`b47a5b4`](https://github.com/ngshiheng/fareview/commit/b47a5b4d40e82a4f51ee58f0cd049cbb8e1cb8b1))

## v1.1.32 (2022-11-18)
### Fix
* **deps:** Update dependency ipython to v8.6.0 ([`fad91c4`](https://github.com/ngshiheng/fareview/commit/fad91c4515bc3b162c5e1561c23f3d580e366149))

## v1.1.31 (2022-11-18)
### Fix
* **deps:** Update dependency sqlalchemy to v1.4.44 ([`b781ee0`](https://github.com/ngshiheng/fareview/commit/b781ee0a5beb81c81382826f038b04e63d1abea9))

## v1.1.30 (2022-11-05)
### Fix
* **spider:** Shopee spider regression due to api requirement change ([`441c28d`](https://github.com/ngshiheng/fareview/commit/441c28de694e7a5488028acba34eeb479defbf23))

## v1.1.29 (2022-11-02)
### Fix
* Downgrade to python3.9 ([`e9895e0`](https://github.com/ngshiheng/fareview/commit/e9895e08508425b3c982cdc1afbbeb8a2303b773))

## v1.1.28 (2022-11-01)
### Fix
* **deps:** Update dependency sqlalchemy to v1.4.42 ([`e5e72cf`](https://github.com/ngshiheng/fareview/commit/e5e72cff0bcd97abf91adadf563838dc031ad98a))

## v1.1.27 (2022-10-31)
### Fix
* **deps:** Update dependency psycopg2-binary to v2.9.5 ([`a471537`](https://github.com/ngshiheng/fareview/commit/a471537bdef3e60b56cbce3144e4bf87b0307939))

## v1.1.26 (2022-10-01)
### Fix
* **deps:** Update dependency sentry-sdk to v1.9.9 ([`9e72c1c`](https://github.com/ngshiheng/fareview/commit/9e72c1cf60c3e8234e11e0cd9e8e4c67e31dcf37))

## v1.1.25 (2022-09-30)
### Fix
* **deps:** Update dependency scrapy to v2.6.3 ([`8e0f22d`](https://github.com/ngshiheng/fareview/commit/8e0f22d12abd20c639ca2179a3baef076eba0a78))

## v1.1.24 (2022-07-28)
### Fix
* **deps:** Update dependency sentry-sdk to v1.9.0 ([`7d7be38`](https://github.com/ngshiheng/fareview/commit/7d7be3808028729938eb94471f9d55ee9583e400))

## v1.1.23 (2022-07-25)
### Fix
* **deps:** Update dependency scrapy to v2.6.2 ([`02f9afb`](https://github.com/ngshiheng/fareview/commit/02f9afb69dc003cf121ca96d677c4a4b1e1fbf5f))

## v1.1.22 (2022-07-21)
### Fix
* **deps:** Update dependency sentry-sdk to v1.8.0 ([`31f350d`](https://github.com/ngshiheng/fareview/commit/31f350d5e00dd88b521c800c7dca5929fdc3587c))

## v1.1.21 (2022-07-15)
### Fix
* **deps:** Update dependency sentry-sdk to v1.7.2 ([`fd5a360`](https://github.com/ngshiheng/fareview/commit/fd5a36091bd84c71fae6a76e4025744233f4e23b))

## v1.1.20 (2022-07-14)
### Fix
* **deps:** Update dependency alembic to v1.8.1 ([`0cc82cf`](https://github.com/ngshiheng/fareview/commit/0cc82cf29956d727224feff9b7a92931e95392da))

## v1.1.19 (2022-07-13)
### Fix
* **deps:** Update dependency sentry-sdk to v1.7.1 ([`78ad07a`](https://github.com/ngshiheng/fareview/commit/78ad07a8fc6f9b96a2d39d9d7e196354cbc3b822))

## v1.1.18 (2022-07-11)
### Fix
* **deps:** Update dependency sentry-sdk to v1.7.0 ([`efad65d`](https://github.com/ngshiheng/fareview/commit/efad65d923f9358ef4008dffcbf6112eb64060f6))

## v1.1.17 (2022-07-11)
### Fix
* **deps:** Update dependency sqlalchemy-utils to v0.38.3 ([`dd33263`](https://github.com/ngshiheng/fareview/commit/dd33263e35814be0544fb7f19303e91cd46c480d))

## v1.1.16 (2022-06-27)
### Fix
* **deps:** Update dependency Scrapy to v2.6.1 ([`0bc9b3d`](https://github.com/ngshiheng/fareview/commit/0bc9b3dcb1642a53e9ce5f7f521003c0d8271a23))

## v1.1.15 (2022-06-24)
### Fix
* **deps:** Update dependency sqlalchemy to v1.4.39 ([`ed77697`](https://github.com/ngshiheng/fareview/commit/ed77697c472aee8cab18f892e652c7674d750e31))

## v1.1.14 (2022-06-22)
### Fix
* **deps:** Update dependency sentry-sdk to v1.6.0 ([`d91c3c0`](https://github.com/ngshiheng/fareview/commit/d91c3c06cab281b62cf14f6aa8094d4ecc8fdf10))
* **scripts:** Fix flag sequence ([`dbaf9c9`](https://github.com/ngshiheng/fareview/commit/dbaf9c9cc888948d1ac62f62a6b27a5645ca304c))

### Documentation
* Add features ([`7084488`](https://github.com/ngshiheng/fareview/commit/708448884e95ec97090bb5bd05209c4fc3a290e4))
* Add badges & table of content ([`6ec19ba`](https://github.com/ngshiheng/fareview/commit/6ec19ba57ef2687d28523dd5c0e856beda249347))

fareview/settings.py
import copy
import os

import scrapy.utils.log
from colorlog import ColoredFormatter

# Logging
color_formatter = ColoredFormatter(
    (
        '%(log_color)s%(levelname)-5s%(reset)s '
        '%(yellow)s[%(asctime)s]%(reset)s'
        '%(white)s %(name)s %(funcName)s %(bold_purple)s:%(lineno)d%(reset)s '
        '%(log_color)s%(message)s%(reset)s'
    ),
    datefmt='%d-%B-%y %H:%M:%S',
    log_colors={
        'DEBUG': 'blue',
        'INFO': 'bold_cyan',
        'WARNING': 'red',
        'ERROR': 'bg_bold_red',
        'CRITICAL': 'red,bg_white',
    },
)

_get_handler = copy.copy(scrapy.utils.log._get_handler)


def _get_handler_custom(*args, **kwargs):
    handler = _get_handler(*args, **kwargs)
    handler.setFormatter(color_formatter)
    return handler


scrapy.utils.log._get_handler = _get_handler_custom


BOT_NAME = 'fareview'
ENVIRONMENT = os.environ.get('ENVIRONMENT', 'development')
SUPPORTED_BRANDS = ['tiger', 'heineken', 'carlsberg', 'guinness', 'asahi']

SPIDER_MODULES = ['fareview.spiders']
NEWSPIDER_MODULE = 'fareview.spiders'

# Scraper API
SCRAPER_API_KEY = os.environ.get('SCRAPER_API_KEY')

# Crawl responsibly by identifying yourself (and your website) on the user-agent
# scrapy-fake-useragent
# https://github.com/alecxe/scrapy-fake-useragent
FAKEUSERAGENT_PROVIDERS = [
    'scrapy_fake_useragent.providers.FakeUserAgentProvider',  # This is the first provider we'll try
    'scrapy_fake_useragent.providers.FakerProvider',  # If FakeUserAgentProvider fails, we'll use faker to generate a user-agent string for us
    'scrapy_fake_useragent.providers.FixedUserAgentProvider',  # Fall back to USER_AGENT value
]

USER_AGENT = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36'

# Obey robots.txt rules
ROBOTSTXT_OBEY = os.environ.get('ROBOTSTXT_OBEY', True)

# Configure maximum concurrent requests performed by Scrapy (default: 16)
CONCURRENT_REQUESTS = 5 if SCRAPER_API_KEY is not None else 16
RETRY_TIMES = os.environ.get('RETRY_TIMES', 10)

# Configure a delay for requests for the same website (default: 0)
# See https://docs.scrapy.org/en/latest/topics/settings.html#download-delay
# See also autothrottle settings and docs
DOWNLOAD_DELAY = 0 if SCRAPER_API_KEY is not None else 2
# The download delay setting will honor only one of:
# CONCURRENT_REQUESTS_PER_DOMAIN = 16
# CONCURRENT_REQUESTS_PER_IP = 16

# Disable cookies (enabled by default)
# COOKIES_ENABLED = False

# Disable Telnet Console (enabled by default)
# TELNETCONSOLE_ENABLED = False

# Override the default request headers:
# DEFAULT_REQUEST_HEADERS = {
#   'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
#   'Accept-Language': 'en',
# }

# Enable or disable spider middlewares
# See https://docs.scrapy.org/en/latest/topics/spider-middleware.html
# SPIDER_MIDDLEWARES = {
#    'fareview.middlewares.FareviewSpiderMiddleware': 543,
# }

# Enable or disable downloader middlewares
# See https://docs.scrapy.org/en/latest/topics/downloader-middleware.html
DOWNLOADER_MIDDLEWARES = {
    'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware': None,
    'scrapy.downloadermiddlewares.retry.RetryMiddleware': None,
    'scrapy_fake_useragent.middleware.RandomUserAgentMiddleware': 400,
    'scrapy_fake_useragent.middleware.RetryUserAgentMiddleware': 401,
    'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware': 810,
}

# Enable or disable extensions
# See https://docs.scrapy.org/en/latest/topics/extensions.html
EXTENSIONS = {
    'fareview.extensions.SentryLogging': -1,
}

# Configure item pipelines
# See https://docs.scrapy.org/en/latest/topics/item-pipeline.html
ITEM_PIPELINES = {
    'fareview.pipelines.ExistingProductPricePipeline': 300,
    'fareview.pipelines.NewProductPricePipeline': 400,
}

# Enable and configure the AutoThrottle extension (disabled by default)
# See https://docs.scrapy.org/en/latest/topics/autothrottle.html
AUTOTHROTTLE_ENABLED = True
# The initial download delay
AUTOTHROTTLE_START_DELAY = 5
# The maximum download delay to be set in case of high latencies
AUTOTHROTTLE_MAX_DELAY = 60
# The average number of requests Scrapy should be sending in parallel to
# each remote server
AUTOTHROTTLE_TARGET_CONCURRENCY = 1.0
# Enable showing throttling stats for every response received:
# AUTOTHROTTLE_DEBUG = False

# Enable and configure HTTP caching (disabled by default)
# See https://docs.scrapy.org/en/latest/topics/downloader-middleware.html#httpcache-middleware-settings
HTTPCACHE_ENABLED = True
HTTPCACHE_EXPIRATION_SECS = os.environ.get('HTTPCACHE_EXPIRATION_SECS', 0)
# HTTPCACHE_DIR = 'httpcache'
# HTTPCACHE_IGNORE_HTTP_CODES = []
# HTTPCACHE_STORAGE = 'scrapy.extensions.httpcache.FilesystemCacheStorage'

# Sentry
# https://stackoverflow.com/questions/25262765/handle-all-exception-in-scrapy-with-sentry
SENTRY_ENABLED = os.environ.get('SENTRY_ENABLED', True)
SENTRY_DSN = os.environ.get('SENTRY_DSN')

# PostgreSQL
DATABASE_CONNECTION_STRING = '{drivername}://{user}:{password}@{host}:{port}/{db_name}'.format(
    drivername='postgresql',
    user=os.environ.get('PG_USERNAME', 'postgres'),
    password=os.environ.get('PG_PASSWORD'),
    host=os.environ.get('PG_HOST', 'localhost'),
    port=os.environ.get('PG_PORT', '5432'),
    db_name=os.environ.get('PG_DATABASE', 'fareview'),
)


# Management Commands
COMMANDS_MODULE = 'fareview.commands'
